src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/MetadataManager.cs

```diff
@@ -765,6 +770,52 @@ internal IEnumerable<NativeLayoutTemplateMethodSignatureVertexNode> GetTemplateM
             return _templateMethodEntries;
         }
 
+        private Dictionary<TypeDesc, TypeDesc> _canonToTemplate;
+
+        public TypeDesc GetTemplateTypeForCanonicalFormType(NodeFactory factory, TypeDesc type)
+        {
+            Debug.Assert(factory.MarkingComplete);
+            Debug.Assert(type.IsCanonicalSubtype(CanonicalFormKind.Specific));
+
+            _canonToTemplate ??= BuildCanonToTemplateLookup(_typesWithConstructedEETypesGenerated);
+            return _canonToTemplate.TryGetValue(type, out TypeDesc result) ? result : type;
+        }
+
+        private Dictionary<TypeDesc, TypeDesc> _necessaryCanonToTemplate;
+
+        public TypeDesc GetTemplateTypeForNecessaryCanonicalFormType(NodeFactory factory, TypeDesc type)
+        {
+            Debug.Assert(factory.MarkingComplete);
+            Debug.Assert(type.IsCanonicalSubtype(CanonicalFormKind.Specific));
+
+            _necessaryCanonToTemplate ??= BuildCanonToTemplateLookup(_typesWithNeccessaryEETypesGenerated);
+            return _necessaryCanonToTemplate.TryGetValue(type, out TypeDesc result) ? result : type;
+        }
+
+        private static Dictionary<TypeDesc, TypeDesc> BuildCanonToTemplateLookup(SortedSet<TypeDesc> e)
+        {
+            var canonTypeToConcreteType = new Dictionary<TypeDesc, List<TypeDesc>>();
+            foreach (TypeDesc constructedType in e)
+            {
+                TypeDesc canonType = constructedType.ConvertToCanonForm(CanonicalFormKind.Specific);
+                if (constructedType != canonType)
+                {
+                    if (!canonTypeToConcreteType.TryGetValue(canonType, out List<TypeDesc> concreteList))
+                        canonTypeToConcreteType.Add(canonType, concreteList = new List<TypeDesc>());
+                    concreteList.Add(constructedType);
+                }
+            }
+
+            var canonToTemplate = new Dictionary<TypeDesc, TypeDesc>();
+            foreach ((TypeDesc canonType, List<TypeDesc> concreteTypes) in canonTypeToConcreteType)
+            {
+                concreteTypes.Sort(TypeSystemComparer.Instance);
```

Comment:
We can just go through the array and pick the smallest element. We do not need to run full Sort.

Reply:
Actually, do we even need to build the list of all candidates? We can just maintain the smallest element in the dictionary as we are populating it.